// fw/lib/util.js
{
    var _ns_ = {
            id: 'fw.lib.macros',
            doc: void 0
        };
}
void 0;
void 0;
void 0;
void 0;
void 0;
void 0;
{
    var _ns_ = {
            id: 'fw.lib.util',
            doc: void 0
        };
}
var isFn = exports.isFn = function isFn(x) {
        return typeof(x) === 'function';
    };
var last = exports.last = function last(arr) {
        return arr.slice(-1).shift();
    };
var isUniq = exports.isUniq = function isUniq(arr) {
        return arr.length === 1;
    };
var filterEmpty = exports.filterEmpty = function filterEmpty(arr) {
        return Array.isArray(arr) ? arr.filter(function (x) {
            return !(x === void 0 || x === null);
        }) : arr;
    };
var once = exports.once = function once(lambda) {
        return function () {
            var callø1 = false;
            return function () {
                var args = Array.prototype.slice.call(arguments, 0);
                return !callø1 ? (function () {
                    callø1 = true;
                    return lambda.apply(void 0, args);
                })() : void 0;
            };
        }.call(this);
    };
