// hu/lib/misc.js
{
    var _ns_ = {
            id: 'hu.lib.macros',
            doc: void 0
        };
    var hu_lib_function = require('./function');
    var curry = hu_lib_function.curry;
    var compose = hu_lib_function.compose;
    var hu_lib_type = require('./type');
    var isString = hu_lib_type.isString;
    var isArray = hu_lib_type.isArray;
    var isNumber = hu_lib_type.isNumber;
    var isObject = hu_lib_type.isObject;
}
void 0;
void 0;
void 0;
void 0;
void 0;
void 0;
void 0;
void 0;
void 0;
void 0;
{
    var _ns_ = {
            id: 'hu.lib.misc',
            doc: void 0
        };
}
var noop = exports.noop = function () {
        return void 0;
    };
var now = exports.now = Date.now;
var log = exports.log = function log() {
        var args = Array.prototype.slice.call(arguments, 0);
        return console ? console.log.apply(console, args) : void 0;
    };
var isBrowser = exports.isBrowser = (function () {
        return typeof(window) === 'object' && window.HTMLElement;
    })();
var _global = exports._global = isBrowser ? window : global;
